### PR TITLE
Support new distribution-validations and add few enhancements to validation workflow

### PR DESCRIPTION
**Pull request by Divyaasm**
### Description
1. Support Zip and Deb validations.
2. Move the download_artifacts method from child classes(except docker) to parent class - validation.py.
3. Changes to validation_args.py to fix the bug https://github.com/opensearch-project/opensearch-build/issues/4441
4. Cluster readiness is checked by requesting the api's instead of adding sleep().
5. Changes to docker-validation:
    1. Add security strong password changes before running the container.
    2. Bring up the services opensearchnode1 and opensearchnode2 only when OS is validated, bringup all the services in 
        compose file when both OS and OSD are provided.
    3. Pull and validate the latest staging docker image in the given version if build number is not provided.
    
Addresses #4424 and #4423

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/validation_workflow/docker/validation_docker.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ def start_cluster(self) -> bool:
     def validation(self) -> bool:
         # STEP 2 . inspect image digest between opensearchproject(downloaded/local) and opensearchstaging(dockerHub)
         if not self.args.using_staging_artifact_only:
-            self.image_names_list = [self.args.OS_image, self.args.OSD_image]
+            self.image_names_list = ['opensearchproject/' + project for project in self.args.projects]
             self.image_names_list = [x for x in self.image_names_list if (os.path.basename(x) in self.args.projects)]
```
Comment (jordarlu) [inline src/validation_workflow/docker/validation_docker.py:65]: what does these two lines of `self.image_name_list` do?

File: src/validation_workflow/deb/validation_deb.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    def installation(self) -> bool:
+        try:
+            for project in self.args.projects:
+                set_password = f' env OPENSEARCH_INITIAL_ADMIN_PASSWORD={get_password(str(self.args.version))}' if project == "opensearch" else ""
+                execute(f'sudo dpkg --purge {project}', ".")
+                execute(f'sudo {set_password} dpkg -i {os.path.basename(self.args.file_path.get(project))}', str(self.tmp_dir.path))
+                time.sleep(80)
+        except:
+            raise Exception("Failed to install Opensearch")
```
Comment (gaiksaya) [inline src/validation_workflow/deb/validation_deb.py:32]: I believe the product might be anything? OpenSearch or Dashboards? Recommending to make the error message generic.

File: src/validation_workflow/rpm/validation_rpm.py [hunk 3]
```diff
@@ -22,26 +21,6 @@ class ValidateRpm(Validation, DownloadUtils):
...
-                else:
-                    self.args.version = self.get_version(self.args.file_path.get(project))
-                    self.check_url(self.args.file_path.get(project))
-            else:
-                if (self.args.artifact_type == "staging"):
-                    self.args.file_path[project] = f"{self.base_url_staging}{project}/{self.args.version}/{self.args.build_number[project]}/linux/{self.args.arch}/{self.args.distribution}/dist/{project}/{project}-{self.args.version}-linux-{self.args.arch}.rpm"  # noqa: E501
-                else:
-                    self.args.file_path[project] = f"{self.base_url_production}{project}/{self.args.version}/{project}-{self.args.version}-linux-{self.args.arch}.rpm"
-                self.check_url(self.args.file_path.get(project))
-        return True
```
Comment (gaiksaya) [inline src/validation_workflow/rpm/validation_rpm.py:44]: Why are we removing the entire RPM method?

File: src/validation_workflow/tar/validation_tar.py [hunk 4]
```diff
@@ -23,29 +22,9 @@ class ValidateTar(Validation, DownloadUtils):
...
-                    self.args.version = self.get_version(self.args.file_path.get(project))
-                    self.check_url(self.args.file_path.get(project))
-            else:
-                if (self.args.artifact_type == "staging"):
-                    self.args.file_path[project] = f"{self.base_url_staging}{project}/{self.args.version}/{self.args.build_number[project]}/linux/{self.args.arch}/{self.args.distribution}/dist/{project}/{project}-{self.args.version}-linux-{self.args.arch}.tar.gz"  # noqa: E501
-                else:
-                    self.args.file_path[project] = f"{self.base_url_production}{project}/{self.args.version}/{project}-{self.args.version}-linux-{self.args.arch}.tar.gz"
-                self.check_url(self.args.file_path.get(project))
-        return True
-
```
Comment (gaiksaya) [inline src/validation_workflow/tar/validation_tar.py:48]: Same question. Why are removing the entire code block?

File: src/validation_workflow/validation_args.py [hunk 5]
```diff
@@ -123,22 +123,30 @@ def __init__(self) -> None:
...
+            if 'opensearch' not in args.file_path.keys() or not set(args.file_path.keys()) <= {'opensearch', 'opensearch-dashboards'}:
+                raise Exception("Missing OpenSearch artifact details! Please provide the valid product names among opensearch and opensearch-dashboards")
             args.distribution = self.get_distribution_type(args.file_path)
             args.projects = args.file_path.keys()
-        if (('opensearch' not in args.projects)):
-            raise Exception("Missing OpenSearch OpenSearch artifact details! Please provide the same along with OpenSearch-Dashboards to validate")
+
+        if args.distribution == "docker":
+            if args.osd_build_number != "latest" and "opensearch-dashboards" not in args.projects:
+                raise Exception("Provide opensearch-dashboards in projects argument to validate OSD")
```
Comment (gaiksaya) [inline src/validation_workflow/validation_args.py:147]: nit:
```suggestion
                raise Exception("Provide opensearch-dashboards in projects argument to validate OpenSearch-Dashboards")
```

**Comment by gaiksaya** [comment]
Hi Divya, can you please add more details in the PR description explaining what changes are being made. Looks like it is doing a lot more than debian and zip validation.
Also please attach the issue number related to this PR change.

Thanks!

File: src/validation_workflow/deb/validation_deb.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+from test_workflow.integ_test.utils import get_password
+from validation_workflow.api_test_cases import ApiTestCases
+from validation_workflow.download_utils import DownloadUtils
+from validation_workflow.validation import Validation
+from validation_workflow.validation_args import ValidationArgs
+
+
+class ValidateDeb(Validation, DownloadUtils):
+    def __init__(self, args: ValidationArgs) -> None:
+        super().__init__(args)
```
Comment (peterzhuamazon) [inline src/validation_workflow/deb/validation_deb.py:22]: Missing tests.

File: src/validation_workflow/deb/validation_deb.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+    def installation(self) -> bool:
+        try:
+            for project in self.args.projects:
+                set_password = f' env OPENSEARCH_INITIAL_ADMIN_PASSWORD={get_password(str(self.args.version))}' if project == "opensearch" else ""
+                execute(f'sudo dpkg --purge {project}', ".")
+                execute(f'sudo {set_password} dpkg -i {os.path.basename(self.args.file_path.get(project))}', str(self.tmp_dir.path))
```
Comment (peterzhuamazon) [inline src/validation_workflow/deb/validation_deb.py:29]: Why using `project` instead of `product`?

File: src/validation_workflow/deb/validation_deb.py (hunk 2 above)
Comment (peterzhuamazon) [reply src/validation_workflow/deb/validation_deb.py:32]: ^^

File: src/validation_workflow/deb/validation_deb.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        except:
+            raise Exception("Failed to install Opensearch")
+        return True
+
+    def start_cluster(self) -> bool:
+        try:
+            for project in self.args.projects:
+                execute(f'sudo systemctl enable {project}', ".")
+                execute(f'sudo systemctl start {project}', ".")
+                time.sleep(20)
```
Comment (peterzhuamazon) [inline src/validation_workflow/deb/validation_deb.py:40]: Why is there a need to sleep 20 since opensearch systemd is using notify type and will wait until success.
Even if you have to wait why not just do a check of API calls every 5sec instead of blindly waiting for 20sec?
Different server spec can change the delay of this startup.

File: src/validation_workflow/docker/validation_docker.py (hunk 1 above)
Comment (peterzhuamazon) [reply src/validation_workflow/docker/validation_docker.py:65]: Second this.

File: src/validation_workflow/docker/validation_docker.py [hunk 9]
```diff
@@ -252,8 +250,9 @@ def run_container(self, image_ids: dict, version: str) -> Any:
         self.replacements = [(f'opensearchproject/{key}:{self.major_version_number}', value) for key, value in image_ids.items()]
 
         list(map(lambda r: self.inplace_change(self.target_yml_file, r[0], r[1]), self.replacements))
-
+        os.environ["OPENSEARCH_INITIAL_ADMIN_PASSWORD"] = get_password(str(version))
         # spin up containers
-        self.docker_compose_up = f'docker-compose -f {self.target_yml_file} up -d'
+        services = "opensearch-node1 opensearch-node2" if "opensearch-dashboards" not in self.args.projects else ""
```
Comment (peterzhuamazon) [inline src/validation_workflow/docker/validation_docker.py:255]: Why are we separating these instead of just spin up the entire docker compose?

File: src/validation_workflow/rpm/validation_rpm.py (hunk 3 above)
Comment (peterzhuamazon) [reply src/validation_workflow/rpm/validation_rpm.py:44]: ^^

File: src/validation_workflow/tar/validation_tar.py (hunk 4 above)
Comment (peterzhuamazon) [reply src/validation_workflow/tar/validation_tar.py:48]: ^^

**Review by peterzhuamazon — CHANGES_REQUESTED** [review]
Hi @Divyaasm ,

As @gaiksaya also pointed out here, the PR is huge with a lot of information but little comments or descriptions on what is happening.

I would suggest you separate this PR into two separate ones, lets review and get deb merge and we understand your approach before going back on win zip, or you can also avoid duplicate changes.

You are also missing tests on the PR. With a big PR like this I am hesitate to just merge through without any tests.

Please take a look at the comments and let us know. 😄 

Thanks.

File: src/validation_workflow/docker/validation_docker.py (hunk 1 above)
Comment (Divyaasm) [reply src/validation_workflow/docker/validation_docker.py:65]: I have unassigned the args self.args.OS_image, self.args.OSD_image in args.py  and declared a list which stores the images names based projects=["opensearch", "opensearchdashboards"]

File: src/validation_workflow/tar/validation_tar.py (hunk 4 above)
Comment (Divyaasm) [reply src/validation_workflow/tar/validation_tar.py:48]: I moved the entire download_artifacts method from child classes(tar, rpm, yum, deb, zip) to parent class - validation.py

**Comment by Divyaasm** [comment]
> Hi Divya, can you please add more details in the PR description explaining what changes are being made. Looks like it is doing a lot more than debian and zip validation. Also please attach the issue number related to this PR change.
> 
> Thanks!

Sure Sayali

File: src/validation_workflow/docker/validation_docker.py (hunk 9 above)
Comment (Divyaasm) [reply src/validation_workflow/docker/validation_docker.py:255]: Added changes to bring up the services opensearchnode1 and opensearchnode2 only when OS is validated, bringup all the services in compose file when both OS and OSD are provided.

File: src/validation_workflow/deb/validation_deb.py (hunk 6 above)
Comment (Divyaasm) [reply src/validation_workflow/deb/validation_deb.py:22]: I'm working on the test cases. Will update it in next commit. Thanks

File: src/validation_workflow/deb/validation_deb.py (hunk 7 above)
Comment (Divyaasm) [reply src/validation_workflow/deb/validation_deb.py:29]: The argument is `project` from the beginning

File: src/validation_workflow/deb/validation_deb.py (hunk 8 above)
Comment (Divyaasm) [reply src/validation_workflow/deb/validation_deb.py]: Implemented this for all the distributions

File: src/validation_workflow/validation.py [hunk 10]
```diff
@@ -53,9 +60,59 @@ def run(self) -> Any:
...
+                self.args.file_path[project] = self.get_filepath(project)
+                self.check_url(self.args.file_path.get(project))
+        return True
+
+    def get_filepath(self, project: str) -> str:
+        file_name_suffix = "tar.gz" if self.args.distribution == "tar" else self.args.distribution
+        if self.args.artifact_type == "staging":
+            if self.args.distribution == "yum":
+                return f"{self.base_url_staging}{project}/{self.args.version}/{self.args.build_number[project]}/{self.args.platform}/{self.args.arch}/rpm/dist/{project}/{project}-{self.args.version}.staging.repo"  # noqa: E501
+            return f"{self.base_url_staging}{project}/{self.args.version}/{self.args.build_number[project]}/{self.args.platform}/{self.args.arch}/{self.args.distribution}/dist/{project}/{project}-{self.args.version}-{self.args.platform}-{self.args.arch}.{file_name_suffix}"  # noqa: E501
```
Comment (peterzhuamazon) [inline src/validation_workflow/validation.py:82]: I think you can still make this multi-lines right?

File: src/validation_workflow/validation_args.py [hunk 11]
```diff
@@ -114,7 +114,7 @@ def __init__(self) -> None:
         parser.add_argument(
             "-f",
             "--force-https",
-            action="store_true",
+            action="store_false",
```
Comment (peterzhuamazon) [inline src/validation_workflow/validation_args.py:117]: Why this change?

File: src/validation_workflow/validation_args.py [hunk 12]
```diff
@@ -123,22 +123,30 @@ def __init__(self) -> None:
...
-            help="(optional) Validate digest only; will not run docker to test API")
+            help="(optional) Validate digest only; will not run docker to test API",
+            dest="validate_digest_only")
         group.add_argument(
             "--using-staging-artifact-only",
             action="store_true",
             default=False,
-            help="(optional) Use only staging artifact to run docker and API test, will not validate digest")
+            help="(optional) Use only staging artifact to run docker and API test, will not validate digest",
+            dest="using_staging_artifact_only")
```
Comment (peterzhuamazon) [inline src/validation_workflow/validation_args.py:133]: Why?
```
dest="using_staging_artifact_only")
```

**Review by peterzhuamazon — COMMENTED** [review]
More questions.

File: src/validation_workflow/validation.py [hunk 13]
```diff
@@ -53,9 +60,59 @@ def run(self) -> Any:
...
+        retry_count = 0
+        while retry_count < max_retry:
+            logging.info(f'sleeping 5sec for retry {retry_count + 1}/{max_retry}')
+            time.sleep(5)
+            if self.check_http_request():
+                logging.info('\n\ncluster is now ready for API test\n\n')
+                return True
+            retry_count += 1
+        logging.error(f"Maximum number of retries ({max_retry}) reached. Cluster is not ready for API test.")
+        return False
```
Comment (peterzhuamazon) [inline src/validation_workflow/validation.py:98]: Cool 👍

File: src/validation_workflow/validation_args.py (hunk 12 above)
Comment (Divyaasm) [reply src/validation_workflow/validation_args.py:133]: I have just added it for easy readability, as the value using-staging-artifact-only is using '-'.

File: src/validation_workflow/docker/validation_docker.py (hunk 1 above)
Comment (jordarlu) [reply src/validation_workflow/docker/validation_docker.py:61]: thanks, Divya, I just feel a bit redundant here becasue you constructed the image_names_list from args.projects on the first line of code, and then you do a filtering from the same args.projects on the next line of code. Could you give me an exmaple of expected output from the 2nd line of code assuming the image_names_list has ["opensearchproject/opensearch", "opensearchproject/opensearchdashboards"]?

File: src/validation_workflow/validation_args.py (hunk 11 above)
Comment (Divyaasm) [reply src/validation_workflow/validation_args.py:117]: Bit confused on this parameter actually. This is always true with current implementation. Now , It is true by default, when -----force-https is passed it will store false.

To remove the confusion we can change the name to --do-not force-https or something similar and assign a value False by default, if provided we check for security plugin and run http/https as required.

Comment (peterzhuamazon) [reply src/validation_workflow/validation_args.py:117]: Why do we keep doing negative params? I still dont understand your point here. If --force-https, then it will always check https, if not / default it will check https or http based on security plugin presence, isnt it?

File: src/validation_workflow/validation_args.py (hunk 12 above)
Comment (peterzhuamazon) [reply src/validation_workflow/validation_args.py:133]: I mean to say why it ended up with `)`? For both dest you added there?

Comment (Divyaasm) [reply src/validation_workflow/validation_args.py:133]: help="(optional) Use only staging artifact to run docker and API test, will not validate digest")  `)` comes from here 
            help="(optional) Use only staging artifact to run docker and API test, will not validate digest",
            dest="using_staging_artifact_only")

File: src/validation_workflow/validation_args.py (hunk 11 above)
Comment (Divyaasm) [reply src/validation_workflow/validation_args.py:117]: I think we want to make the validation workflow confined to `https` all the time. Unless an external argument  --force-https is provided we do not want to validate without security plugin(i.e http)
Adding @gaiksaya

Comment (gaiksaya) [reply src/validation_workflow/validation_args.py:117]: Right! Security is always assumed to be enabled by default especially for validation.

File: src/validation_workflow/docker/validation_docker.py (hunk 1 above)
Comment (Divyaasm) [reply src/validation_workflow/docker/validation_docker.py:61]: I get it now. Sorry missed this line while editing. It's of no use. Will remove it Thanks for pointing.

File: src/validation_workflow/validation_args.py (hunk 12 above)
Comment (peterzhuamazon) [reply src/validation_workflow/validation_args.py:133]: Then correct the location of the `)`?

File: src/validation_workflow/validation_args.py (hunk 11 above)
Comment (Divyaasm) [reply src/validation_workflow/validation_args.py:117]: Hi @peterzhuamazon, addressed all the changes except this. Can I know your thoughts?

Comment (peterzhuamazon) [reply src/validation_workflow/validation_args.py:117]: After talking to @Divyaasm as there are some confusion.

The param was meant to `disable` https-only check, aka allow http if security plugin not present.
See this flow:

validation force-https drawio

Therefore, the initial naming of `--force-https` is not applicable anymore.

Suggest using `--allow-http` so that, if such parameter not present it will always force to check https no matter what; else, if the parameter present it will allow for `http` check if security plugin is not present.

Thanks.

File: src/validation_workflow/validation_args.py [hunk 14]
```diff
@@ -123,22 +123,32 @@ def __init__(self) -> None:
...
+            if 'opensearch' not in args.file_path.keys() or not set(args.file_path.keys()) <= {'opensearch', 'opensearch-dashboards'}:
+                raise Exception("Missing OpenSearch artifact details! Please provide the valid product names among opensearch and opensearch-dashboards")
             args.distribution = self.get_distribution_type(args.file_path)
             args.projects = args.file_path.keys()
-        if (('opensearch' not in args.projects)):
-            raise Exception("Missing OpenSearch OpenSearch artifact details! Please provide the same along with OpenSearch-Dashboards to validate")
+
+        if args.distribution == "docker":
+            if args.osd_build_number != "latest" and "opensearch-dashboards" not in args.projects:
+                raise Exception("Provide opensearch-dashboards in projects argument to validate OpenSearch-Dashboards")
```
Comment (peterzhuamazon) [inline src/validation_workflow/validation_args.py:149]: Probably change this msg to be much more clear such as:
```
You cannot provide a osd_build_number without specifying opensearch-dashboards in --projectsProbably change this msg to be much more clear such as:
```
You cannot provide a osd_build_number without specifying opensearch-dashboards in --projects
```

Or later we can just skip osd test if user provided osd build number without specifying projects as osd.
```

Or later we can just skip osd test if user provided osd build number without specifying projects as osd.

File: src/validation_workflow/validation_args.py [hunk 15]
```diff
@@ -123,22 +123,32 @@ def __init__(self) -> None:
...
+            if not(args.validate_digest_only or args.using_staging_artifact_only):
```
Comment (peterzhuamazon) [inline src/validation_workflow/validation_args.py:150]: Nit: `if not(A) and not(B)` or a better condition for understanding.
We will raise exception if both of these params are empty.

File: src/validation_workflow/validation_args.py [hunk 16]
```diff
@@ -146,14 +146,14 @@ def __init__(self) -> None:
 
         if args.distribution == "docker":
             if args.osd_build_number != "latest" and "opensearch-dashboards" not in args.projects:
-                raise Exception("Provide opensearch-dashboards in projects argument to validate OpenSearch-Dashboards")
-            if not(args.validate_digest_only or args.using_staging_artifact_only):
+                raise Exception("osd_build_number argument cannot be provided without specifying opensearch-dashboards in --projects")
```
Comment (peterzhuamazon) [inline src/validation_workflow/validation_args.py:149]: Probably should use `--osd-build-number`

File: src/validation_workflow/validation_args.py (hunk 11 above)
Comment (peterzhuamazon) [reply src/validation_workflow/validation_args.py]: Resolved now.

File: src/validation_workflow/validation.py [hunk 17]
```diff
@@ -53,9 +60,66 @@ def run(self) -> Any:
...
+        return f"{self.base_url_production}{project}/{self.args.version}/{project}-{self.args.version}-{self.args.platform}-{self.args.arch}.{file_name_suffix}"
+
+    def check_cluster_readiness(self) -> bool:
+        max_retry = 20
+        retry_count = 0
+        while retry_count < max_retry:
+            logging.info(f'sleeping 5sec for retry {retry_count + 1}/{max_retry}')
+            time.sleep(5)
+            if self.check_http_request():
+                logging.info('\n\ncluster is now ready for API test\n\n')
```
[the remark this answers is not in the document]
Comment (peterzhuamazon) [reply src/validation_workflow/validation.py]: Resolved now.

File: src/validation_workflow/validation_args.py (hunk 15 above)
Comment (peterzhuamazon) [reply src/validation_workflow/validation_args.py]: Resolved now.

**Review by peterzhuamazon — APPROVED** [review]

**Comment by peterzhuamazon** [comment]
Thanks @Divyaasm for addressing these comments.
Approved now.

Thanks.

File: src/validation_workflow/validation_args.py (hunk 16 above)
Comment (peterzhuamazon) [reply src/validation_workflow/validation_args.py]: Thanks.

**Review by jordarlu — APPROVED** [review]

**Comment by peterzhuamazon** [comment]
Hi @Divyaasm you need to address the python test failure.
